### PR TITLE
Memoize routes and named routes

### PR DESCRIPTION
**Pull request by jtoar**
In `RouterImpl`, `routes` and `namedRoutes` are recalculated every render, but their values only change if `Router`'s `children` change, which&mdash;on the fly&mdash;it never does. 

Wrapping `routes` and `namedRoutes` in React 's [`useMemo`](https://reactjs.org/docs/hooks-reference.html#usememo) hook removes the hassle of recalculating these two values every render.

> NOTE: The way `namedRoutes` is defined, it technically does remember if it was called since it closes over a boolean that it sets, but this is maybe a more React-ish way of doing it?

### CODE REVIEW

File: packages/router/src/router.js [hunk 1]
```diff
@@ -105,30 +105,38 @@ const RouterImpl = ({
...
-        return React.Children.toArray(children).map((route) =>
-          React.cloneElement(route, {
-            private: true,
-            unauthenticatedRedirect: unauthenticated,
-            role: role,
-          })
-        )
-      })
-      .reduce((a, b) => a.concat(b), []) || []
+  console.log('router impl rendered')
```
Comment (jtoar) [inline packages/router/src/router.js:108]:
```suggestion
```

File: packages/router/src/router.js [hunk 2]
```diff
@@ -105,30 +105,38 @@ const RouterImpl = ({
...
+
```
Comment (jtoar) [inline packages/router/src/router.js:109]: Remove embarrassing console log
```suggestion
```

**Comment by Tobbe** [comment]
Optimizations are a fun topic 😃 Did you do before/after measurements on both memory and execution time?

**Comment by jtoar** [comment]
@Tobbe I did not! I didn't even think of it as an optimization though it is. :sweat_smile: How would I go about measuring those dimensions? I imagine that this PR wouldn't really make anything more performant until you had a lot of routes.

But I think it does make the code more readable in a way: it signals more explicitly that we know these values depend on `children` and only need to change if children does.

This also might be superseded at some point? I feel like both `routes` and `namedroutes` can be computed in advance or at built-time or something.

**Comment by Tobbe** [comment]
> How would I go about measuring those dimensions?

I was hoping for you to be the teacher here 😄 So far React has always been fast enough for my needs, so I haven't had to do any profiling. 

When hooks first came out I jumped on the hype-train and added `useCallback` everywhere. But then I realized it actually just made the code more verbose, for no good reason, and also actually probably made it slower in a lot of cases... This is a good blog post on the subject https://kentcdodds.com/blog/usememo-and-usecallback

**Comment by jtoar** [comment]
@Tobbe Totally! I got into profiling while diagnosing the layout rerendering&mdash;more concerned about losing state than speed. State-wise, it doesn't really matter if we recalculate the routes every time, it just seemed like an obvious thing that didn't need to be done on every navigation.

I'll look into gathering metrics though!

**Review by peterp — APPROVED** [review]
LGTM!

**Pull request merged**